torch/distributed/distributed_c10d.py

```diff
@@ -2972,18 +2975,61 @@ def new_group(ranks=None, timeout=default_pg_timeout, backend=None, pg_options=N
     Returns:
         A handle of distributed group that can be given to collective calls.
     """
+    return _create_new_group_internal(ranks, timeout, backend, pg_options, global_naming=True)
+
+
+def _get_backend_from_str(backend: Optional[str] = None) -> Backend:
+    # Default to the same backend as the global process group
+    # if the backend is not specified.
+    if not backend:
+        default_pg = _get_default_group()
+        default_backend, _ = _pg_map[default_pg]
+        backend = default_backend
+    return Backend(backend)
+
+
+def new_sub_group(ranks=None, timeout=default_pg_timeout, backend=None, pg_options=None):
+    """
+    Creates a new distributed subgroup in MPMD style.
+
+    This function behaves in the same way as ``new_group`` except that ths function must
+    be called only from the ranks set in ``ranks`` instead of all ranks with ``new_group``.
+
+    N.B. MPI backend is not supported.
+    """
+    # MPI backend doesn't have a Store, so we can't handle new_sub_group
+    if _get_backend_from_str(backend) == Backend.MPI:
+        raise RuntimeError("MPI backend doesn't support new_sub_group")
+    if ranks is not None and get_rank() not in ranks:
+        raise RuntimeError("new_sub_group can only be called from a member rank")
+    return _create_new_group_internal(ranks, timeout, backend, pg_options, global_naming=False)
+
+
+def _process_group_name(ranks, use_sequential):
+    if use_sequential:
+        global _group_count
+        pg_name = str(_group_count)
+        _group_count += 1
+    else:
+        global _pg_names
+        pg_name = hashlib.sha1(bytes("_".join(map(str, ranks)), "utf-8")).hexdigest()
```

Comment:
should we prefer sha2 vs sha1 hashes?

Reply:
It doesn't matter much the hash function we use.
We're using the hash function as a deterministic PRNG.
SHA2 buys us 64 extra bits of randomness which I find hard to believe it would be needed. SHA1 produces a chance of collision of 10^-33 for 100 billion process groups.

The crucial issue here is ensuring that it's not possible to construct the same input string to the hashing function for two distinct set of ranks.